### DISCUSSION TITLE
Convert legacy EVP_PKEY to non-legacy

### DISCUSSION THREAD

**Discussion by beldmit** [openssl/openssl#22458] [Q&A]
I have a EVP_PKEY created via `EVP_PKEY_new`/`EVP_PKEY_set1_RSA` (or for `EVP_PKEY_set1_EC_KEY`). Is there a simple way to convert it into non-legacy key in case when it doesn't have custom RSA_METHOD?

**Comment by t8m** [comment]
You would have to do todata/fromdata parameters export import IMO.

**Reply by beldmit** [reply]
From `EVP_PKEY_todata` manual page:

`These functions only work with key management methods coming from a provider.`

And keymgmt is NULL in that case

**Reply by t8m** [reply]
Hmm... I thought we have the ctrl conversion layer that should handle that. Or it works only on individual get_BN_param calls?

**Comment by beldmit** [comment]
As I mentioned, I am rewriting openssh-portable to EVP API. The first step is migrating sshkey structure to EVP_PKEY for RSA/EC.

For compatibility purposes I use EVP_PKEY_set1_RSA/EC_KEY in most cases.

The PR I added https://github.com/openssh/openssh-portable/pull/453 does the trick but I came across the issue with the key conversion:
```
cp regress/ssh-rsa_oldfmt
./ssh-keygen -p -N '' -m PEM -f regress/ssh-rsa_oldfmt
```
fails when I try to use the 3.0 API.

I had to comment out the "new" code path (https://github.com/openssh/openssh-portable/pull/453/commits/5b360418cf727fc3821a2236f1c5a8c0f5aaa362) to make this command work.

The debugger shows that the mgmt pointer is NULL, and I understand why :)
So no encoder is found. 

Is this behavior intentional? What is the best way to fix it?

**Reply by levitte** [reply]
Ah, OSSL_ENCODER stuff only works with provider based keys, that's true.  However, you can still call the PEM functions with all sorts of keys.  Furthermore, I had a radical suggestion in the openssh PR that should help you get past this limitation.

**Reply by beldmit** [reply]
Is it the only known limitation? E.g. will "generic" RSA being assigned this way use FIPS provider for crypto?

**Reply by levitte** [reply]
I frankly don't remember...  I know for a fact that all cryptographic functions (signing, verifying, that sort of thing) are designed to convert things under the hood, i.e they perform an export/import dance from legacy to provider based.

Ah, there's another limitation: the export from legacy keys depend on having our implementations.  If openssh has its own `EVP_PKEY_ASN1_METHOD` instances, then you're out of luck.  This is due to a few new functions that help doing this, that were added in OpenSSL 3.0, but there are no public functions to help set such functions.

**Reply by beldmit** [reply]
There are no `EVP_PKEY_ASN1_METHOD` instances in the OpenSSH code - just EC_KEY_METHOD/RSA_METHOD for PKCS#11 purposes

**Reply by levitte** [reply]
Ok, then I don't expect any problem of that sort
